layer1/executor/tasks/tests/snapshot_test.go

```diff
@@ -0,0 +1,450 @@
+		require.Nil(t, err)
+		signer := &crypto.BNGroupSigner{}
+		err = signer.SetPrivk(dkgState.GroupPrivateKey.Bytes())
+		require.Nil(t, err)
+		bnSigners = append(bnSigners, signer)
+		groupKey, err := bn256.MarshalBigIntSlice(dkgState.MasterPublicKey[:])
+		if err != nil {
+			t.Fatal(err)
+		}
+		err = signer.SetGroupPubk(groupKey)
+		require.Nil(t, err)
+	}
+
+	// Valid at 1024
+	dkgState, err := dkgState.GetDkgState(suite.DKGStatesDbs[0])
+	require.Nil(t, err)
+	grpSig1024, bClaimsBin1024, err := GenerateSnapshotData(1337, 1024, bnSigners, n, dkgState.MasterPublicKey[:], false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	bclaims := &objs.BClaims{}
+	err = bclaims.UnmarshalBinary(bClaimsBin1024)
+	require.Nil(t, err)
+
+	bh := &objs.BlockHeader{
+		BClaims:  bclaims,
+		TxHshLst: [][]byte{},
+		SigGroup: grpSig1024,
+	}
+
+	currentHeight, err := eth.GetCurrentHeight(ctx)
+	assert.Nil(t, err)
+
+	for idx := 0; idx < n; idx++ {
+		snapshotState := &state.SnapshotState{
+			Account:     eth.GetDefaultAccount(),
+			BlockHeader: bh,
+		}
+
+		err := state.SaveSnapshotState(suite.DKGStatesDbs[idx], snapshotState)
+		require.Nil(t, err)
+	}
+
+	for idx := 0; idx < n; idx++ {
+		snapshotTask := snapshots.NewSnapshotTask(currentHeight, n, idx)
+
+		err := snapshotTask.Initialize(ctx, nil, suite.DKGStatesDbs[idx], fixture.Logger, suite.Eth, fixture.Contracts, "snapshotTask", "task-id", nil)
+		assert.Nil(t, err)
+		err = snapshotTask.Prepare(ctx)
+		assert.Nil(t, err)
+
+		snapshotState, err := state.GetSnapshotState(snapshotTask.GetDB())
+		assert.Nil(t, err)
+
+		shouldExecute, err := snapshotTask.ShouldExecute(ctx)
+		if shouldExecute {
+			txn, taskError := snapshotTask.Execute(ctx)
+			amILeading, err := utils.AmILeading(
+				suite.Eth,
+				ctx,
+				fixture.Logger,
+				snapshotState.LastSnapshotHeight,
+				snapshotState.RandomSeedHash,
+				snapshotTask.NumOfValidators,
+				snapshotTask.ValidatorIndex,
+				snapshotState.DesperationFactor,
+				snapshotState.DesperationDelay,
+			)
+			assert.Nil(t, err)
+			if amILeading {
+				assert.Nil(t, taskError)
+				rcptResponse, err := fixture.Watcher.Subscribe(ctx, txn, nil)
+				assert.Nil(t, err)
+				tests.WaitGroupReceipts(t, suite.Eth, []transaction.ReceiptResponse{rcptResponse})
+			} else {
+				assert.Nil(t, txn)
+				require.NotNil(t, taskError)
+				assert.True(t, taskError.IsRecoverable())
+			}
+		}
+	}
+}
+
+func Test_LeaderElectionWithRandomData(t *testing.T) {
+	valMin := 4
+	valMax := 50
+	fixture := setupEthereum(t, 5)
+	eth := fixture.Client
+	ctx := context.Background()
+	for i := 0; i < 100; i++ {
+		valNum := getRandomNumberInRange(valMin, valMax)
+		valIndex := getRandomNumberInRange(0, valNum)
+		hash := []byte(uuid.New().String())
+		keccakedHash := ethCrypto.Keccak256(hash)
+		var hashSlice32 [32]byte
+		copy(hashSlice32[:], keccakedHash)
+
+		currentHeight, err := eth.GetCurrentHeight(ctx)
+		assert.Nil(t, err)
+
+		blocksToMine := getRandomNumberInRange(1, 1000)
```

Comment:
It _should_ be ok since we're doing ~100 rounds of this, but we need to be careful with randomness in tests. It makes the tests non-deterministic which makes debugging locally a pain if something goes wrong in CI and can sometimes cause fluctuations in code coverage results which get flagged (as happened with the badger trie).

Reply:
I would prefer to use a deterministic source with a seed for the randomness. This gives us a google random sample that may be re-played in other places.

Reply:
Since Im rolling this 100 times I can use i as the seed.